attrd/commands.c

```diff
@@ -748,6 +754,68 @@ attrd_lookup_or_create_value(GHashTable *values, const char *host, xmlNode *xml)
     return(v);
 }
 
+static gboolean
+check_current_attribute_from_sync_response(crm_node_t *peer, xmlNode *xml, const char *search_host, const char *name, const char *current_value)
+{
+    xmlNode *child = NULL;
+    gboolean find = FALSE;
+
+    for (child = __xml_first_child(xml); child != NULL; child = __xml_next(child)) {
```

Comment:
I'm concerned about having to loop through this entire list for each attribute the local node knows about. Normally, it won't be much, but as we scale to hundreds of nodes and hundreds of resources (with fail count attributes), it could become significant.

What do you think of this approach:
* Add a new "seen" flag to attribute_value_t
* When getting ATTRD_OP_SYNC_RESPONSE, set seen=FALSE for all known attribute values before processing
* In attrd_peer_update(), set seen=TRUE once "v" is known
* Then we can omit this function, and simply check v->seen==FALSE

Reply:
I will examine the method you presented.

Reply:
Hi Ken,

How about with the correction like the following?

```
@attrd/internal.h
(snip)
typedef struct attribute_value_s {
        uint32_t nodeid;
        gboolean is_remote;
        char *nodename;
        char *current;
        char *requested;
        gboolean seen;
} attribute_value_t;
(snip)

@attrd/commands.c
(snip)
static void
clear_attribute_value_seen(void)
{
    GHashTableIter aIter;
    GHashTableIter vIter;
    attribute_t *a;
    attribute_value_t *v = NULL;

    g_hash_table_iter_init(&aIter, attributes);
    while (g_hash_table_iter_next(&aIter, NULL, (gpointer *) & a)) {
        g_hash_table_iter_init(&vIter, a->values);
        while (g_hash_table_iter_next(&vIter, NULL, (gpointer *) & v)) {
            v->seen = FALSE;
            crm_debug("Clear seen flag %s[%s] = %s.", a->id, v->nodename, v->current);
        }
    }
}
(snip)

void
attrd_peer_message(crm_node_t *peer, xmlNode *xml)
{
(snip)
    } else if (safe_str_eq(op, ATTRD_OP_SYNC_RESPONSE)
              && safe_str_neq(peer->uname, attrd_cluster->uname)) {
        xmlNode *child = NULL;
        crm_info("Processing %s from %s", op, peer->uname);

        /* Clear the seen flag for attribute processing held only in the own node. */
        clear_attribute_value_seen();

        for (child = __xml_first_child(xml); child != NULL; child = __xml_next(child)) {
            host = crm_element_value(child, F_ATTRD_HOST);
            attrd_peer_update(peer, child, host, TRUE);
        }

        if (peer_state == election_won) {
            /* Synchronize if there is an attribute held only by own node that Writer does not have. */
            attrd_current_only_attribute_update(peer, xml);
        }
    }

(snip)
void
attrd_current_only_attribute_update(crm_node_t *peer, xmlNode *xml)
{
(snip)
    g_hash_table_iter_init(&aIter, attributes);
    while (g_hash_table_iter_next(&aIter, NULL, (gpointer *) & a)) {
        g_hash_table_iter_init(&vIter, a->values);
        while (g_hash_table_iter_next(&vIter, NULL, (gpointer *) & v)) {
            if (safe_str_eq(v->nodename, attrd_cluster->uname) && v->seen == FALSE) {
                if (check_current_attribute_from_sync_response(peer, xml, attrd_cluster->uname, a->id, v->current) == FALSE) {
                    crm_debug("Syncing %s[%s] = %s to everyone.(from local only attributes)", a->id, v->nodename, v->current);
(snip)

void
attrd_peer_update(crm_node_t *peer, xmlNode *xml, const char *host, bool filter)
{
(snip)
    } else {
        crm_trace("Unchanged %s[%s] from %s is %s", attr, host, peer->uname, value);
    }

    /* Set the seen flag for attribute processing held only in the own node. */                                                                            
    v->seen = TRUE;

    /* If this is a cluster node whose node ID we are learning, remember it */
    if ((v->nodeid == 0) && (v->is_remote == FALSE)
(snip)
```

Reply:
Yes, that looks good!

Reply:
Okay!
Thank you for confirmation!